text/0018-open-context-provider-framework.md

```diff
@@ -0,0 +1,157 @@
+---
+feature name: Open context provider framwork
+start date: 2020-05-27
+rfc pr: 
+related issue: https://github.com/aws/aws-cdk-rfcs/issues/18
+---
+
+# Summary
+
+The CDK has a super-handy feature in the form of the ContextProvider ( https://docs.aws.amazon.com/cdk/latest/guide/context.html ). These
+ContextProviders are currently hard-coded into the CDK CLI, and additional ContextProviders cannot be added by external sources. This
+proposal is to make the ContextProvider mechanism pluggable so that 3rd parties can create ContextProviders without adding them to the CDK
+repository, and users of the CDK can use these 3rd party ContextProviders.
+
+# Motivation
+
+Development within the CDK Construct library itself benefits greatly from this capability. When a new ContextProvider is needed, then a CDK
+contributor can just create it. Unfortunately, since the providers are hard-coded into the CDK, no 3rd party construct library can create their
+own ContextProvider; other than by contributing it to the CDK, of course. Some ContextProviders may be too specialized for reasonable consideration
+as a CDK contribution.
+
+Instead 3rd party libraries are required to implement workarounds, like CfnCustomResources that resolve during deployment instead of during
+synthesis. This increases the size of the Cfn template generated by the app, increases the time required to deploy the stack, adds unnecessary
+extra resources to the user's account, and makes it possible for a value generated by the CfnCustomResource to differ from one deployment
+to the next.
+
+Unfortunately, these CfnCustomResources required by 3rd party construct libraries may also be a security risk to the user. The lambda that
```

Comment:
This smells a little like FUD. While I personally would have preferred a way to restrict Lambdas Invoke permissions to CloudFormation (which is afaik not possible, and when I asked for it they told me to write the custom resource to do a callback to the Stack and check whether the resource was currently `_IN_PROGRESS, ... https://tt.amazon.com/0150388604), this should not be the main factor in deciding between custom resource and context provider.